deploy to openshift, got Error: Attempt to unlock async@*, which hasn't been locked

git push openshift master, then got this error.

``` bash

remote:   - Installing dependencies w/ new version of npm ...
remote:
remote: npm info it worked if it ends with ok
remote: npm info using npm@1.4.3
remote: npm info using node@v0.10.26
remote: npm info preinstall cantas@1.0.0
remote: npm ERR! Error: Attempt to unlock connect-redis@1.4.x, which hasn't been locked
remote: npm ERR!     at unlock (/var/lib/openshift/536769dd500446ba9c00014f/app-root/data/node-v0.10.26-linux-x64/lib/node_modules/npm/lib/cache.js:1304:11)
remote: npm ERR!     at cb (/var/lib/openshift/536769dd500446ba9c00014f/app-root/data/node-v0.10.26-linux-x64/lib/node_modules/npm/lib/cache.js:646:5)
remote: npm ERR!     at /var/lib/openshift/536769dd500446ba9c00014f/app-root/data/node-v0.10.26-linux-x64/lib/node_modules/npm/lib/cache.js:655:20
remote: npm ERR!     at /var/lib/openshift/536769dd500446ba9c00014f/app-root/data/node-v0.10.26-linux-x64/lib/node_modules/npm/lib/cache.js:1282:20
remote: npm ERR!     at afterMkdir (/var/lib/openshift/536769dd500446ba9c00014f/app-root/data/node-v0.10.26-linux-x64/lib/node_modules/npm/lib/cache.js:1013:14)
remote: npm ERR!     at /var/lib/openshift/536769dd500446ba9c00014f/app-root/data/node-v0.10.26-linux-x64/lib/node_modules/npm/node_modules/mkdirp/index.js:37:53
remote: npm ERR!     at Object.oncomplete (fs.js:107:15)
remote: npm ERR! If you need help, you may report this *entire* log,
remote: npm ERR! including the npm and node versions, at:
remote: npm ERR!     <http://github.com/npm/npm/issues>

....
```

--- a/.openshift/lib/utils
+++ b/.openshift/lib/utils
@@ -4,26 +4,11 @@
 #
 
 
-#  Returns the configured Node version
+#  Returns the configured Node version - defaults to 0.8.24.
 function get_node_version() {
-   # read the app's package.json file
-   # package_json="${OPENSHIFT_REPO_DIR}package.json"
-
-   # attempt to detect the desired node.js version
-   # engine=$(grep "\"node\"" "$package_json" | tail -n 1 | sed -e 's/^.*" *[~=<>]*[=<>]* *v*\([0-9][^" ]*\) *[^"]*".*$/\1/p;d')
-
-   # assume that a deprecated marker value is a reasonable default
-   # marker="${OPENSHIFT_REPO_DIR}.openshift/markers/NODEJS_VERSION"
-   # marker_ver=$(egrep -v "^\s*#.*" "$marker" 2>/dev/null | egrep -v "^\s*$" | tail -1)
-
-   # use the OpenShift cart default of 0.6.20 as a final option:
-   # if [ -z $engine ]; then
-   #     echo ${marker_ver-"0.6.20"}
-   # else
-   #     echo ${engine}
-   # fi
-
-   echo "0.10.26"
+   marker="$OPENSHIFT_REPO_DIR/.openshift/markers/NODEJS_VERSION"
+   nodejs_ver=$(egrep -v "^\s*#.*" "$marker" | egrep -v "^\s*$" | tail -1)
+   echo "${nodejs_ver:-"0.8.24"}"
 
 }  #  End of function  get_node_version.
 
@@ -38,7 +23,7 @@ function get_node_install_dir() {
 #  Returns the path to the npm binary.
 function get_npm_bin_path() {
    ver=${1:-"$(get_node_version)"}
-   echo "$(get_node_install_dir)node-v$ver-linux-x64/bin"
+   echo "$(get_node_install_dir)/node-v$ver-linux-x64/bin"
 
 }  #  End of function  get_npm_bin_path.
 
@@ -69,8 +54,8 @@ function _install_nodejs() {
    ver=${1:-"$(get_node_version)"}
 
    # Sample download links:
-   #    http://nodejs.org/dist/v0.8.9/node-v0.8.9-linux-x64.tar.gz
-   #    http://nodejs.org/dist/v0.9.1/node-v0.9.1-linux-x64.tar.gz
+   #    http://nodejs.org/dist/v0.8.24/node-v0.8.24-linux-x64.tar.gz
+   #    http://nodejs.org/dist/v0.10.10/node-v0.10.10-linux-x64.tar.gz
    zfile="node-v${ver}-linux-x64.tar.gz"
    zlink="http://nodejs.org/dist/v${ver}/${zfile}"
 
@@ -93,6 +78,26 @@ function _install_nodejs() {
 
 }  #  End of function  _install_nodejs.
 
+
+#  Ensure npm and node symlinks exist in ~/.node_modules/.bin/ directory.
+function _ensure_symlinks_in_node_modules_bin_dir() {
+   zdir="$HOME/.node_modules/.bin/"
+   if [ -d "$zdir" ]; then
+      ln -sf "$(get_npm_bin_path)/npm"   "$zdir"
+      ln -sf "$(get_node_bin_path)/node" "$zdir"
+   fi
+
+}  #  End of function  _ensure_symlinks_in_node_modules_bin_dir.
+
+
+#  Ensure context - this is a bit too hacky.
+function _ensure_context_exists() {
+   zenv="$OPENSHIFT_HOMEDIR/nodejs/configuration/node.env"
+   if ! egrep '^\s*function\s+nodejs_context' "$zenv" > /dev/null 2>&1; then
+      echo -e "function nodejs_context() { bash -c \"\$@\"; }\n" >> "$zenv" || :
+   fi
+
+}  #  End of function  _ensure_context_exists.
 
 
 #  Ensure the shell env setup bits are added to user's .bash_profile.
@@ -121,8 +126,8 @@ function ensure_node_is_installed() {
 
    #
    #  To re-download and reinstall Node.js, uncomment these lines.
-        # rm -f downloads/node-v${ver}-linux-x64.tar.gz
-        # rm -rf $OPENSHIFT_DATA_DIR/node-v${ver}-linux-x64/
+   #     rm -f downloads/node-v${ver}-linux-x64.tar.gz
+   #     rm -rf $OPENSHIFT_DATA_DIR/node-v${ver}-linux-x64/
    #
    #  Note:  This function could be run multiple times on every git push,
    #         so use w/ caution (do once).
@@ -133,6 +138,12 @@ function ensure_node_is_installed() {
    else
       _install_nodejs "$ver"
    fi
+
+   #  Ensure node and npm symlinks exist in ~/.node_modules/.bin directory.
+   _ensure_symlinks_in_node_modules_bin_dir
+
+   #  Ensure context exists.
+   _ensure_context_exists
 
    #  Ensure .bash_profile sets up path for custom Node[.js] version.
    _ensure_bash_profile_setup

--- a/.openshift/markers/NODEJS_VERSION
+++ b/.openshift/markers/NODEJS_VERSION
@@ -1,0 +1,11 @@
+#  Uncomment one of the version lines to select the node version to use.
+#  The last "non-blank" version line is the one picked up by the code in
+#  .openshift/lib/utils
+#  Default: 0.10.25
+#
+#  0.8.24
+#  0.9.1
+#  0.10.25
+#  0.11.11
+0.10.28
+

